### ISSUE TITLE
CloudStack allows deploying VMs on Networks that should not be accessed by VM's owner

### ISSUE THREAD

**Issue by GabrielBrascher** [apache/cloudstack#6049]
##### ISSUE TYPE

 * Bug Report

##### COMPONENT NAME

~~~
API
~~~

##### CLOUDSTACK VERSION

~~~
(probably affects also versions before 4.15, but not tested before)
4.15
4.16
main (4.17)
~~~

##### CONFIGURATION

Advanced network with security groups

##### OS / ENVIRONMENT

N/A

##### SUMMARY

VMs can be deployed with Networks that have a domain ID where the VM’s owner / domain should not be able to.

Even with the user not being able to list the UUID, it is still possible to use the API command to deploy when passing the UUID.
This also can be easily reproduced when the root admin (which can list all networks) uses a "private" network when deploying a VM to a user that does not have access to the respective "private" / "restricted" network.

This can become a very serious issue in case this grows into multiple VMs on the wrong networks.
Also, this is a security issue.

##### STEPS TO REPRODUCE

~~~
Deploy VM for an account that has no access to the network ID
~~~

##### EXPECTED RESULTS

~~~
VM is not deployed (propper error message should be thrown)
~~~

##### ACTUAL RESULTS

~~~
VM is deployed on network that the user / domain should not be able to
~~~

**Comment by PaulAngus** [comment]
Can you do this by using the UUID as a normal user from an account that should not be able to see this network?

**Comment by GabrielBrascher** [comment]
@PaulAngus yes, it can be a normal User calling the API and the VM will be deployed on any network, regardless of the owner/domain of the network.

Normal users are unable to list these networks but they can still deploy VMs there.

**Comment by weizhouapache** [comment]
@GabrielBrascher 
I have fixed a similar issue but with L2 network in #5784 

what type of network do you use ?

**Comment by GabrielBrascher** [comment]
Thanks for raising that PR, @weizhouapache.

I have not dug deeper into your PR, so I might be missing another context beyond the lines you presented here.
Please, correct me if I am missing something regarding your PR.

Unfortunately, I don't see how changing from "caller" to "vmOwner" in the `_accountMgr.checkAccess,` does fix this issue.

I've tested both ways:
1. Root admin deploying VM for a user account
2. the user account itself deploying a VM (thus, **caller = vmOwner**)

Both ways the VM was deployed even with the owner / caller not holding permissions to deploy VMs on the respective network.

**Comment by weizhouapache** [comment]
> Thanks for raising that PR, @weizhouapache.
> 
> I have not dug deeper into your PR, so I might be missing another context beyond the lines you presented here. Please, correct me if I am missing something regarding your PR.
> 
> Unfortunately, I don't see how changing from "caller" to "vmOwner" in the `_accountMgr.checkAccess,` does fix this issue.
> 
> I've tested both ways:
> 
> 1. Root admin deploying VM for a user account
> 2. the user account itself deploying a VM (thus, **caller = vmOwner**)
> 
> Both ways the VM was deployed even with the owner / caller not holding permissions to deploy VMs on the respective network.

@GabrielBrascher 
OK, could you share your cloudstack version, zone type and network type ?

in my PR #5769, there are many changes related to network permissoins, mostly in server/src/main/java/com/cloud/network/NetworkModelImpl.java 

However, first of all, we need to get a consensus if this is a critical issue for 4.16.1.0.

**Comment by rohityadavcloud** [comment]
Since the normal user can't list a network or has knowledge of the UUID then in 99.99% of cases this doesn't sound like a critical issue. This still however is a bug that should be fixed, I think it's upto the 4.16.1 RM @sureshanaparti to decide/triage.

@GabrielBrascher I also wonder if there are other APIs that have similar behaviour where a normal user can't list the resource or has knowledge of UUID, but other APIs could allow accepting those uuids for an operation on a resource.

**Comment by GabrielBrascher** [comment]
@rohityadavcloud I was thinking the same. It is a good thing to run some tests with other API endpoints as well.
